File: LovePlay-ASDK/LovePlay/Class/Zone/Controller/ZoneListViewController.m
//
//  ZoneListViewController.m
//  LovePlay
//
//  Created by Yuns on 2017/1/29.
//  Copyright © 2017年 Yuns. All rights reserved.
//

#import "ZoneListViewController.h"
//M
#import "ZoneListModel.h"
//V
#import "ZoneListCell.h"
#import "ZoneListHeaderView.h"
//C
#import "DiscussListViewController.h"

static NSString *zoneListCell = @"zoneListCell";
static NSString *zoneListHeader = @"zoneListHeader";

@interface ZoneListViewController ()<UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout>
//UI
@property (nonatomic, strong) UICollectionView *colletionView;
//Data
@property (nonatomic, strong) NSArray *zoneDatas;
@end

@implementation ZoneListViewController

#pragma mark - life cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    [self addCollectionView];
    [self loadData];
}

#pragma mark - init
- (void)addCollectionView
{
    [self.view addSubview:self.colletionView];
}

- (void)loadData
{
    [[HttpRequest sharedInstance] GET:ZoneDiscussURL parameters:nil success:^(id response) {
        NSLog(@"list-succ : %@", response);
        ZoneListModel *listModel = [ZoneListModel modelWithJSON:response[@"info"]];
        _zoneDatas = listModel.discuzList;
        [_colletionView reloadData];
    } failure:^(NSError *error) {
        NSLog(@"list-fail : %@", error);
    }];
}

#pragma mark - collectionView dataSource
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return _zoneDatas.count;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    ZoneDiscussDetail *discussDetail = _zoneDatas[section];
    return discussDetail.detailList.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    ZoneDiscussDetail *discussDetail = _zoneDatas[indexPath.section];
    ZoneDiscussItem *discussItem = discussDetail.detailList[indexPath.row];
    ZoneListCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:zoneListCell forIndexPath:indexPath];
    cell.discussItem = discussItem;
    return cell;
}

#pragma mark + header and footer
- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
{
    ZoneDiscussDetail *discussDetail = _zoneDatas[indexPath.section];
    if ([kind isEqualToString:UICollectionElementKindSectionHeader]) {
        ZoneListHeaderView *headerView = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:zoneListHeader forIndexPath:indexPath];
        headerView.titleName = discussDetail.type.typeName;
        return headerView;
    }
    return nil;
}

#pragma mark - collectionView delegate
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    ZoneDiscussDetail *discussDetail = _zoneDatas[indexPath.section];
    ZoneDiscussItem *discussItem = discussDetail.detailList[indexPath.row];
    DiscussListViewController *discussListViewController = [[DiscussListViewController alloc] init];
    discussListViewController.fid = @(discussItem.fid).stringValue;
    [self.navigationController pushViewController:discussListViewController animated:YES];
}

#pragma mark + header and footer
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section
{
    return CGSizeMake(self.view.bounds.size.width, 30);
}

#pragma mark - setter / getter
- (UICollectionView *)colletionView
{
    if (!_colletionView) {
        UICollectionViewFlowLayout *flowLayout = [[UICollectionViewFlowLayout alloc] init];
//        flowLayout.sectionHeadersPinToVisibleBounds = YES;
        flowLayout.itemSize = CGSizeMake((self.view.bounds.size.width - 1) / 2, 70);
        flowLayout.minimumLineSpacing = 1;
        flowLayout.minimumInteritemSpacing = 1;
        UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:self.view.bounds collectionViewLayout:flowLayout];
        collectionView.backgroundColor = RGB(245, 245, 245);
        collectionView.dataSource = self;
        collectionView.delegate = self;
        [collectionView registerClass:[ZoneListCell class] forCellWithReuseIdentifier:zoneListCell];
        [collectionView registerClass:[ZoneListHeaderView class] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:zoneListHeader];
        _colletionView = collectionView;
    }
    return _colletionView;
}

#pragma mark - other
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: LovePlay-Swift3/LovePlay/Class/Zone/Controller/ZoneListViewController.swift
//
//  ZoneListViewController.swift
//  LovePlay-UIKit-Swift
//
//  Created by weiying on 2017/3/10.
//  Copyright © 2017年 yuns. All rights reserved.
//

import UIKit
import Alamofire

class ZoneListViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {

    var zoneDatas : [ZoneListModel]? = [ZoneListModel]()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.addSubViews()
        self.loadData()
    }

    // MARK: - private
    private func addSubViews() {
        self.view.addSubview(self.colletionView)
    }
    
    private func loadData() {
        let urlStr = BaseURL + ZoneDiscussURL
        Alamofire.request(urlStr).responseJSON { (response) in
            print(response)
            switch response.result.isSuccess {
            case true :
                if let value = response.result.value as? NSDictionary {
                    let info = value["info"] as? NSDictionary
                    let discuzList = info?["discuzList"] as? [NSDictionary]
                    for detailList in discuzList! {
                        let listModel = ZoneListModel.deserialize(from: detailList)
                        self.zoneDatas?.append(listModel!)
                    }
                    
                    self.colletionView.reloadData()
                }
            case false :
                print(response.result.error!)
            }
        }
    }
    
    // MARK: - collectionView dataSource
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        if (self.zoneDatas?.isEmpty)! {
            return 0
        }
        return (self.zoneDatas?.count)!
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if let detailList = self.zoneDatas?[section].detailList {
            return detailList.count
        }
        return 0
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = ZoneListCell.cellWithCollectionView(collectionView: collectionView, indexPath: indexPath)
        let detailList = self.zoneDatas?[indexPath.section].detailList
        let discussItem = detailList?[indexPath.row]
        cell.listModel = discussItem
        return cell
    }
    
    // MARK: - collectionView delegate
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        
        if kind == UICollectionElementKindSectionHeader {
            let listModel = self.zoneDatas?[indexPath.section]
            let sectionHeader : ZoneListSectionHeader = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: NSStringFromClass(ZoneListSectionHeader.self), for: indexPath) as! ZoneListSectionHeader
            sectionHeader.titleTextLabel.text = listModel?.type?.typeName
            return sectionHeader
        }
        return UICollectionReusableView()
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
        return CGSize(width: self.view.bounds.size.width, height: 30)
    }
    
    // MARK: - setter / getter
    lazy var colletionView : UICollectionView = {
        let flowLayout = UICollectionViewFlowLayout()
        flowLayout.itemSize = CGSize(width: (self.view.bounds.size.width - 1) / 2, height: 70.0)
        flowLayout.minimumLineSpacing = 1
        flowLayout.minimumInteritemSpacing = 1
        let colletionView : UICollectionView = UICollectionView(frame: self.view.bounds, collectionViewLayout: flowLayout)
        colletionView.backgroundColor = RGB(r: 245, g: 245, b: 245)
        colletionView.delegate = self
        colletionView.dataSource = self
        colletionView.alwaysBounceVertical = true
        colletionView.register(ZoneListCell.self, forCellWithReuseIdentifier: NSStringFromClass(ZoneListCell.self))
        colletionView.register(ZoneListSectionHeader.self, forSupplementaryViewOfKind:UICollectionElementKindSectionHeader, withReuseIdentifier: NSStringFromClass(ZoneListSectionHeader.self))
        return colletionView
    }()
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

}
